contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+
+####  Refs
+ Ring has no refs.
+
+####  Object
+ Ring will not have objects. (only contracts) 
+
+### Pattern matching
+Pattern matching is probably outside the first iteration of Ring, but
+we definitely want it in the final language.
+
+### Builtins
+#### Events
+Ring has events which logs a structured message to the contract log in
+the resulting blockchain transaction.
+
+```
+event(name: string, value: any)
+```
+
+#### Transactions
+Ring can generate blockchain transactions.
+```
+transaction(tx_type: transaction, argument: record )
+```
+
+#### State
+
+A Ring contract has a state. You must declare the state record type in
+the contract. You can then update the state through mutable record
+update.
+
+```
+type state = { mutable field: string
+             , mutable value: uint};
+let state = {field: “”, value : 0};
+let store_value = (f: string, v: uint) => {
+    state.field = f;
+    state.value = v;
+}
+```
+#### Contract primitives
+```
+creator() : () => address
+```
+The address (pubkey) of the entity signing for the contract creation transaction
+```
+caller() : () => address
+```
+The address of the entity currently calling the contract.
+```
+value() : () => uint
+```
+The amount of coins transferred to the contract when calling.
+```
+balance() : () => uint
+```
+The amount of coins currently in the contract account
+```
+contract_address() : () => address;
+```
+The address of the contract.
+
+Block references
+```
+block_height() : () => uint;    /* Current block number */
+block_timestamp() : () => uint; /* Block timestamp */
+```
+
+
+### Exception
+There is one exception: abort (reason:string)
+
+### JSX
+Ring does not support inline html syntax.
+
+## The lifetime of a contract
+
+# Killing a contract
+
+There is no selfdestruct instruction in the aevm as in the Ethereum
+Virtual Machine instead there is a disable transaction which the
+creator of a contract can issue. When a contract is disabled no new
+contract can call the old contract.
+
+When a contract is posted to the chain all references to other
+contracts are checked and a reference counter in each contract is
+increased. You can only post a contract to the chain if all the
+contracts referred to are enabled.
+
+When a contract is disabled all other contracts it refer to get their
+reference count decreased.
+
+If a contract is disabled and its reference count is zero a miner can
+choose to garbage collect the contract.
+
+The reference count of a contract is handled as the account balance
+and kept in the state tree of the miner and the merkle hash is
+included in the state hash in each block just as with balances.
+
+The transaction for creating a contract has an extra fee called
+deposit which has to be an even number. The disable transaction is
+free but the miner and the creator get half of the deposit fee each at
+contract disable thus encouraging creators to disable their contracts
+and miners to pick disable transactions.
```

Comment:
Wouldn't a miner have an inherent incentive to pick up disable transactions since that means they're going to have to potentially store/process less data?
And would the size of this deposit be mandated by the network or freely chosen?

Reply:
The inherent incentive would be very weak though. If it's more immediately profitable to pick other transactions, a miner wouldn't pick disable transactions until they're starting to get low on memory/disk.

I'm more concerned about malicious disabling of contracts. If disabling a contract is a separate transaction that seems difficult to prevent. Wouldn't it make more sense to only allow the contract code to disable itself? At least that way a user of a contract has the ability to check under which conditions the contract can be disabled.

Reply:
Yes, this is a valid point (and the reason the disable transaction is listed as "to be decided").

@knarz What is your view on having a disable transaction or only a disable instruction in the VM?